libcaf_core/caf/async/fwd.hpp

```diff
@@ -39,4 +41,15 @@ class spsc_buffer;
 
 using execution_context_ptr = intrusive_ptr<execution_context>;
 
+// -- type traits --------------------------------------------------------------
+
+template <class T>
+struct is_publisher : std::false_type {};
+
+template <class T>
+struct is_publisher<publisher<T>> : std::true_type {};
+
+template <class T>
+inline constexpr bool is_publisher_v = is_publisher<T>::value;
+
```

Comment:
Not fond of putting implementation in fwd headers, just for header hygiene. There's one more comment below.